File: Trees/427. Construct Quad Tree/Solution.py
from typing import List

# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':
        # N : no of rows and columns
        # Time Complexity: O(N * N * logN)
        # Space Complexity: O(logN)
        n = len(grid)
        def isSame(n, r, c):
            for i in range(n):
                for j in range(n):
                    if grid[r][c] == grid[r + i][c + j]:
                        continue
                    return False
            return True

        def dfs(n, r, c):
            if isSame(n, r, c):
                return Node(grid[r][c], True)

            n //= 2
            topLeft = dfs(n, r, c)
            topRight = dfs(n, r, c + n)
            bottomLeft = dfs(n, r + n, c)
            bottomRight = dfs(n, r + n, c + n)
            return Node(0, False, topLeft, topRight, bottomLeft, bottomRight)
        return dfs(n, 0, 0)

File: Trees/427. Construct Quad Tree/Solution.java
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}


class Solution {
    private Boolean isSame(int n, int r, int c, int[][] grid) {
        for (int i = 0; i < n; i++) 
            for (int j = 0; j < n; j++) 
                if (grid[r][c] != grid[r + i][c + j]) return false;
        return true;
    }

    private Node dfs(int n, int r, int c, int[][] grid) {
        if (isSame(n, r, c, grid)) return new Node(grid[r][c] == 1, true);

        n /= 2;
        Node topLeft = dfs(n, r, c, grid);
        Node topRight = dfs(n, r, c + n, grid);
        Node bottomLeft = dfs(n, r + n, c, grid);
        Node bottomRight = dfs(n, r + n, c + n, grid);
        
        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }

    public Node construct(int[][] grid) {
        // N : no of rows and columns
        // Time Complexity: O(N * N * logN)
        // Space Complexity: O(logN)
        int n = grid.length;
        return dfs(n, 0, 0, grid);
    }
}